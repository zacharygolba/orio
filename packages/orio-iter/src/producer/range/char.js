// @flow

import * as result from 'orio-result'
import { AsIterator, ToString } from 'orio-traits'
import type { Producer } from 'orio-types'

import NumberRange from './number'

export const MIN_CHAR: string = String.fromCodePoint(0)
export const MAX_CHAR: string = String.fromCodePoint(0x10ffff)

@ToString
@AsIterator
export default class CharRange implements Producer<string> {
  /*:: @@iterator: () => Iterator<string> */
  source: NumberRange

  constructor(start?: string = MIN_CHAR, end?: string = MAX_CHAR) {
    this.source = new NumberRange(start.codePointAt(0), end.codePointAt(0))
  }

  drop(): void {
    this.source.drop()
  }

  next(): IteratorResult<string, void> {
    const next = this.source.next()

    if (next.done) {
      return next
    }

    return result.next(String.fromCodePoint(next.value))
  }
}
