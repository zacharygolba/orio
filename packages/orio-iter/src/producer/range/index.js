// @flow

export { default as CharRange } from './char'
export { default as NumberRange } from './number'
