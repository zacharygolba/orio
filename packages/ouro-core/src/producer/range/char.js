// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'

import type { Producer } from '../../types'

import NumberRange from './number'

export const MIN_CHAR: string = String.fromCodePoint(0)
export const MAX_CHAR: string = String.fromCodePoint(0x10ffff)

@ToString
@AsIterator
export default class CharRange implements Producer<string> {
  /*:: @@iterator: () => Iterator<string> */
  source: NumberRange
  size: number

  constructor(start?: string = MIN_CHAR, end?: string = MAX_CHAR) {
    this.source = new NumberRange(start.codePointAt(0), end.codePointAt(0))
    this.size = this.source.size
  }

  drop(): void {
    this.source.drop()
  }

  next(): IteratorResult<string, void> {
    const next = this.source.next()

    if (next.done) {
      return next
    }

    return result.next(String.fromCodePoint(next.value))
  }
}
