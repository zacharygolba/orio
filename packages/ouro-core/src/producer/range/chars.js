// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'

import Numbers from './numbers'

export const MIN_CHAR: string = String.fromCodePoint(0)
export const MAX_CHAR: string = String.fromCodePoint(0x10ffff)

@ToString
@AsIterator
export default class Chars implements Iterator<string> {
  /*:: @@iterator: () => Iterator<string> */
  source: Numbers
  size: number

  constructor(start?: string = MIN_CHAR, end?: string = MAX_CHAR) {
    this.source = new Numbers(start.codePointAt(0), end.codePointAt(0))
    this.size = this.source.size
  }

  next(): IteratorResult<string, void> {
    const next = this.source.next()

    if (next.done) {
      return next
    }

    return result.next(String.fromCodePoint(next.value))
  }
}
