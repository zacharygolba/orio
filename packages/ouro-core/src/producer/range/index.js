// @flow

export { default as Chars } from './chars'
export { default as Numbers } from './numbers'
