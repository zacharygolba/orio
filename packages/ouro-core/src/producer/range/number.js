// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'

import type { Producer } from '../../types'

@ToString
@AsIterator
export default class Numbers implements Producer<number> {
  /*:: @@iterator: () => Iterator<number> */
  done: boolean
  end: number
  start: number
  state: number
  step: number

  constructor(s?: number = 0, e?: number = Infinity) {
    const start = +s
    const end = +e

    if (Number.isNaN(start) || Number.isNaN(end)) {
      this.drop()
    } else if (start > end) {
      this.end = end
      this.start = Number.isFinite(start) ? start : Number.MAX_SAFE_INTEGER
      this.step = -1
      this.state = this.start
    } else {
      this.end = Number.isFinite(end) ? end : Number.MAX_SAFE_INTEGER
      this.start = start
      this.step = 1
      this.state = this.start
    }
  }

  drop(): void {
    this.done = true
    this.end = 0
    this.start = 0
    this.state = 0
    this.step = 0
  }

  next(): IteratorResult<number, void> {
    if (this.done) {
      return result.done()
    }

    const value = this.state

    if (value === this.end) {
      this.done = true
    } else {
      this.state += this.step
    }

    return result.next(value)
  }
}
