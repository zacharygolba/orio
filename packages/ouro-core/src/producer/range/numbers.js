// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'

@ToString
@AsIterator
export default class Numbers implements Iterator<number> {
  /*:: @@iterator: () => Iterator<number> */
  done: boolean
  end: number
  size: number
  start: number
  state: number
  step: number

  constructor(s?: number = 0, e?: number = Infinity) {
    const start = Number(s)
    const end = Number(e)

    this.size = Math.abs(start - end) + 1

    if (Number.isNaN(start) || Number.isNaN(end) || this.size === 0) {
      this.done = true
      this.end = 0
      this.start = 0
      this.step = 0
    } else if (start > end) {
      this.end = end
      this.start = Number.isFinite(start) ? start : Number.MAX_SAFE_INTEGER
      this.step = -1
    } else {
      this.end = Number.isFinite(end) ? end : Number.MAX_SAFE_INTEGER
      this.start = start
      this.step = 1
    }

    this.state = this.start
  }

  next(): IteratorResult<number, void> {
    if (this.done) {
      return result.done()
    }

    const value = this.state

    if (value === this.end) {
      this.done = true
    } else {
      this.state += this.step
    }

    return result.next(value)
  }
}
