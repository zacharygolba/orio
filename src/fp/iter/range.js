// @flow

export default function* range(start?: number = 0, end?: number = Infinity): Iterator<number> {
  for (let next = start; next < end; next += 1) {
    yield next
  }
}
