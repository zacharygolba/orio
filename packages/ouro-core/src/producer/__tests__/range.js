// @flow

import { CharProducer, NumberProducer } from '../range'

describe('NumberProducer', () => {
  let producer

  describe('->', () => {
    beforeEach(() => {
      producer = new NumberProducer(1, 3)
    })

    test('#constructor()', () => {
      expect(new NumberProducer()).toMatchSnapshot()
      expect(new NumberProducer(1)).toMatchSnapshot()
      expect(new NumberProducer(1, 3)).toMatchSnapshot()
      expect(new NumberProducer(1, Infinity)).toMatchSnapshot()
      // $ExpectError
      expect(new NumberProducer('dog', 'cat')).toMatchSnapshot()
    })

    test('#@@iterator()', () => {
      for (const item of producer) {
        expect(item).toMatchSnapshot()
      }
    })

    test('#next()', () => {
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
    })
  })

  describe('<-', () => {
    beforeEach(() => {
      producer = new NumberProducer(3, 1)
    })

    test('#constructor()', () => {
      expect(new NumberProducer(3, 1)).toMatchSnapshot()
      expect(new NumberProducer(Infinity, 1)).toMatchSnapshot()
    })

    test('#@@iterator()', () => {
      for (const item of producer) {
        expect(item).toMatchSnapshot()
      }
    })

    test('#next()', () => {
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
    })
  })
})

describe('CharProducer', () => {
  let producer

  describe('->', () => {
    beforeEach(() => {
      producer = new CharProducer('a', 'c')
    })

    test('#@@iterator()', () => {
      for (const item of producer) {
        expect(item).toMatchSnapshot()
      }
    })

    test('#next()', () => {
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
    })
  })

  describe('<-', () => {
    beforeEach(() => {
      producer = new CharProducer('c', 'a')
    })

    test('#constructor()', () => {
      expect(producer).toMatchSnapshot()
      expect(new CharProducer()).toMatchSnapshot()
    })

    test('#@@iterator()', () => {
      for (const item of producer) {
        expect(item).toMatchSnapshot()
      }
    })

    test('#next()', () => {
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
      expect(producer.next()).toMatchSnapshot()
    })
  })
})
