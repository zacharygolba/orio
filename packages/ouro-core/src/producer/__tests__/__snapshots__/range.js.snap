// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharRange <- #constructor() 1`] = `
CharRange {
  "source": Numbers {
    "end": 97,
    "start": 99,
    "state": 99,
    "step": -1,
  },
}
`;

exports[`CharRange <- #constructor() 2`] = `
CharRange {
  "source": Numbers {
    "end": 1114111,
    "start": 0,
    "state": 0,
    "step": 1,
  },
}
`;

exports[`CharRange <- #drop() 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`CharRange <- #drop() 2`] = `
CharRange {
  "source": Numbers {
    "done": true,
    "end": 0,
    "start": 0,
    "state": 0,
    "step": 0,
  },
}
`;

exports[`CharRange <- #next() 1`] = `
Object {
  "done": false,
  "value": "c",
}
`;

exports[`CharRange <- #next() 2`] = `
Object {
  "done": false,
  "value": "b",
}
`;

exports[`CharRange <- #next() 3`] = `
Object {
  "done": false,
  "value": "a",
}
`;

exports[`CharRange <- #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`CharRange -> #drop() 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`CharRange -> #drop() 2`] = `
CharRange {
  "source": Numbers {
    "done": true,
    "end": 0,
    "start": 0,
    "state": 0,
    "step": 0,
  },
}
`;

exports[`CharRange -> #next() 1`] = `
Object {
  "done": false,
  "value": "a",
}
`;

exports[`CharRange -> #next() 2`] = `
Object {
  "done": false,
  "value": "b",
}
`;

exports[`CharRange -> #next() 3`] = `
Object {
  "done": false,
  "value": "c",
}
`;

exports[`CharRange -> #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`NumberRange <- #constructor() 1`] = `
Numbers {
  "end": 1,
  "start": 3,
  "state": 3,
  "step": -1,
}
`;

exports[`NumberRange <- #constructor() 2`] = `
Numbers {
  "end": 1,
  "start": 9007199254740991,
  "state": 9007199254740991,
  "step": -1,
}
`;

exports[`NumberRange <- #drop() 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`NumberRange <- #drop() 2`] = `
Numbers {
  "done": true,
  "end": 0,
  "start": 0,
  "state": 0,
  "step": 0,
}
`;

exports[`NumberRange <- #next() 1`] = `
Object {
  "done": false,
  "value": 3,
}
`;

exports[`NumberRange <- #next() 2`] = `
Object {
  "done": false,
  "value": 2,
}
`;

exports[`NumberRange <- #next() 3`] = `
Object {
  "done": false,
  "value": 1,
}
`;

exports[`NumberRange <- #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`NumberRange -> #constructor() 1`] = `
Numbers {
  "end": 9007199254740991,
  "start": 0,
  "state": 0,
  "step": 1,
}
`;

exports[`NumberRange -> #constructor() 2`] = `
Numbers {
  "end": 9007199254740991,
  "start": 1,
  "state": 1,
  "step": 1,
}
`;

exports[`NumberRange -> #constructor() 3`] = `
Numbers {
  "end": 3,
  "start": 1,
  "state": 1,
  "step": 1,
}
`;

exports[`NumberRange -> #constructor() 4`] = `
Numbers {
  "end": 9007199254740991,
  "start": 1,
  "state": 1,
  "step": 1,
}
`;

exports[`NumberRange -> #constructor() 5`] = `
Numbers {
  "done": true,
  "end": 0,
  "start": 0,
  "state": 0,
  "step": 0,
}
`;

exports[`NumberRange -> #drop() 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`NumberRange -> #drop() 2`] = `
Numbers {
  "done": true,
  "end": 0,
  "start": 0,
  "state": 0,
  "step": 0,
}
`;

exports[`NumberRange -> #next() 1`] = `
Object {
  "done": false,
  "value": 1,
}
`;

exports[`NumberRange -> #next() 2`] = `
Object {
  "done": false,
  "value": 2,
}
`;

exports[`NumberRange -> #next() 3`] = `
Object {
  "done": false,
  "value": 3,
}
`;

exports[`NumberRange -> #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;
